### ISSUE TITLE
Undefined references when installing with Mingw-w64

### ISSUE THREAD

**Issue by Ketho** [lunarmodules/luasocket#337]
I could not install LuaSocket with the original MinGW project, having the same issue as https://github.com/diegonehab/luasocket/issues/241. So I switched to Mingw-w64 where it was suggested to use x86_64 as architecture and sj_lj for exception model which solved that particular error. I'm renaming the `gcc.exe` to `mingw32-gcc.exe` to make it work for LuaRocks.

But the Mingw-w64 `x86_64-posix-sjlj-rev0` or `i686-posix-sjlj-rev0` configs don't compile for me and give undefined reference errors.

Version:
- Lua: [lua-5.3.6_Win32_bin](http://luabinaries.sourceforge.net/download.html), lua-5.3.6_Win32_dllw6_lib
- LuaRocks: [luarocks-3.7.0-windows-32](https://github.com/luarocks/luarocks/wiki/Download)
- Mingw-w64: 8.1.0
- Windows 10: 21H1

Logs:
- [x86_64](https://gist.github.com/Ketho/ae1f78e97c53926d0aac2e2fc384b2ad#file-luasocket-w64-x86_64-txt)
```
C:/Program Files/mingw-w64/x86_64-8.1.0-posix-sjlj-rt_v6-rev0/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/../../../../x86_64-w64-mingw32/lib/../lib/dllcrt2.o:crtdll.c:(.text+0x16d): undefined reference to `_execute_onexit_table'
C:/Program Files/mingw-w64/x86_64-8.1.0-posix-sjlj-rt_v6-rev0/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/../../../../x86_64-w64-mingw32/lib/../lib/dllcrt2.o:crtdll.c:(.text+0x8): undefined reference to `_initialize_onexit_table'
C:/Program Files/mingw-w64/x86_64-8.1.0-posix-sjlj-rt_v6-rev0/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/../../../../x86_64-w64-mingw32/lib/../lib/dllcrt2.o:crtdll.c:(.text+0x38b): undefined reference to `_register_onexit_function'
src/luasocket.o:luasocket.c:(.text+0x1f): undefined reference to `luaL_checkinteger'
src/luasocket.o:luasocket.c:(.text+0x2a): undefined reference to `lua_gettop'
```

- [i686](https://gist.github.com/Ketho/ae1f78e97c53926d0aac2e2fc384b2ad#file-luasocket-w64-i686-txt)
```
C:/Program Files (x86)/mingw-w64/i686-8.1.0-posix-sjlj-rt_v6-rev0/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.1.0/../../../../i686-w64-mingw32/lib/../lib/dllcrt2.o:crtdll.c:(.text+0xb): undefined reference to `_initialize_onexit_table'
C:/Program Files (x86)/mingw-w64/i686-8.1.0-posix-sjlj-rt_v6-rev0/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.1.0/../../../../i686-w64-mingw32/lib/../lib/dllcrt2.o:crtdll.c:(.text+0x59): undefined reference to `_imp__Sleep@4'
C:/Program Files (x86)/mingw-w64/i686-8.1.0-posix-sjlj-rt_v6-rev0/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.1.0/../../../../i686-w64-mingw32/lib/../lib/dllcrt2.o:crtdll.c:(.text+0xb6): undefined reference to `_imp__Sleep@4'
C:/Program Files (x86)/mingw-w64/i686-8.1.0-posix-sjlj-rt_v6-rev0/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.1.0/../../../../i686-w64-mingw32/lib/../lib/dllcrt2.o:crtdll.c:(.text+0x168): undefined reference to `_execute_onexit_table'
C:/Program Files (x86)/mingw-w64/i686-8.1.0-posix-sjlj-rt_v6-rev0/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.1.0/../../../../i686-w64-mingw32/lib/../lib/dllcrt2.o:crtdll.c:(.text+0x3d3): undefined reference to `_register_onexit_function'
src/timeout.o:timeout.c:(.text+0x41): undefined reference to `_imp__Sleep@4'
src/timeout.o:timeout.c:(.text+0x9c): undefined reference to `_imp__GetSystemTimeAsFileTime@4'
```

**Comment by daniilpetrovdp** [comment]
This luarocks program is very unbalanved for Windows, as I can see

**Comment by Ketho** [comment]
I had to use e.g. cl.exe from [Build Tools for Visual Studio 2019](https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2019) to make it work.

![image](https://user-images.githubusercontent.com/1073877/140643381-eaa6ffbf-6fec-4a59-9973-dea9502cfca9.png)

It really is difficult to compile luasocket on windows ![image](https://cdn.discordapp.com/emojis/238539575264280578.png?size=16)

**Comment by b97tsk** [comment]
#296 #308 #321 #336

Everybody looks here: https://github.com/lunarmodules/luasocket/issues/316#issuecomment-987675268.

**Comment by ALENTL** [comment]
> 

How did you just do that. Can you help me also do that. I know your discord id is Ketho#9716 but I can't msg you there
![image](https://user-images.githubusercontent.com/82425585/178788006-92cc63ea-9d7e-4619-b952-07bbf15f8752.png)
I have been constantly posting this error on Lua Scripter's discord server. Please help me my friend

**Comment by alerque** [comment]
@ALENTL Your issues is the same as #321, not what this issue is tracking. Please see the latest release which should fix it.

**Comment by alerque** [comment]
Several people have contributed Windows related fixes that wend into the current release, v3.1.0. I do not have access to Windows myself to test any of this, but I would appreciate if anyone here could report whether those fixes cover this issue or not, and if not of course I am open to PR contributions from anyone that can fix this issue.

**Comment by alerque** [comment]
Thanks @Ketho, that's quite helpful to know.
